deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl

```diff
@@ -1217,3 +1513,504 @@ maybe_apply_default_topic_exchange(Exchange, _DefaultTopicExchange) ->
+        end.
+
+delete_queue(QRes, Username) ->
+    case rabbit_amqqueue:with(
+           QRes,
+           fun (Q) ->
+                   io:format("Delete queue ~p~n", [rabbit_queue_type:delete(Q, false, false, Username)])
+           end,
+           fun (not_found) ->
+                   ok;
+               ({absent, Q, crashed}) ->
+                   rabbit_classic_queue:delete_crashed(Q, Username);
+               ({absent, Q, stopped}) ->
+                   rabbit_classic_queue:delete_crashed(Q, Username);
+               ({absent, _Q, _Reason}) ->
+                   ok
+           end) of
+        {ok, _N} ->
+            ok;
+        ok ->
+            ok
+    end.
+
+ensure_binding(#resource{name = QueueBin}, {"", Queue}, _State) ->
+    %% i.e., we should only be asked to bind to the default exchange a
+    %% queue with its own name
+    QueueBin = list_to_binary(Queue),
+    ok;
+ensure_binding(QName, {Exchange, RoutingKey}, _State = #proc_state{auth_login = Username,
+                                                                   vhost = VHost}) ->
+    Binding = #binding{source = rabbit_misc:r(VHost, exchange, list_to_binary(Exchange)),
+                       destination = QName,
+                       key = list_to_binary(RoutingKey)},
+    case rabbit_binding:add(Binding, Username) of
+        {error, {resources_missing, [{not_found, Name} | _]}} ->
+            rabbit_amqqueue:not_found(Name);
+        {error, {resources_missing, [{absent, Q, Reason} | _]}} ->
+            rabbit_amqqueue:absent(Q, Reason);
+        {error, {binding_invalid, Fmt, Args}} ->
+            rabbit_misc:protocol_error(precondition_failed, Fmt, Args);
+        {error, #amqp_error{} = Error} ->
+            rabbit_misc:protocol_error(Error);
+        ok ->
+            ok
+    end.
+
+check_resource_access(User, Resource, Perm, Context) ->
+    V = {Resource, Context, Perm},
+    Cache = case get(permission_cache) of
+                undefined -> [];
+                Other     -> Other
+            end,
+    case lists:member(V, Cache) of
+        true ->
+            ok;
+        false ->
+            rabbit_access_control:check_resource_access(User, Resource, Perm, Context),
+            CacheTail = lists:sublist(Cache, ?MAX_PERMISSION_CACHE_SIZE-1),
+            put(permission_cache, [V | CacheTail]),
+            ok
+    end.
+
+do_native_login(Creds, Heartbeat, Version,  State = #proc_state{peer_addr = Addr,
+                                                                vhost = VHost}) ->
+    {Username, AuthProps} = case Creds of
+                                {ssl, {Username0, none}}-> {Username0, []};
+                                {_, {Username0, Password}} -> {Username0, [{password, Password},
+                                                                           {vhost, VHost}]}
+                            end,
+
+    maybe
+        {ok, User} ?= rabbit_access_control:check_user_login(Username, AuthProps),
```

Comment:
`rabbit_core_metrics:auth_attempt_failed()` is missing if user login is refused.